### PR TITLE
Add callback support

### LINKED ISSUE #9: Support calling and being called by Python and HOC functions, etc

### LINKED ISSUE #11: FInitializeHandler
Initialization is a critical aspect of defining a simulation.

We must allow a way for MATLAB functions to control initialization. In Python and HOC this is done using the [`FInitializeHandler`](https://nrn.readthedocs.io/en/latest/python/simctrl/programmatic.html#finitializehandler) class which registers callbacks at four separate stages of initialization.

### PR DESCRIPTION
**Pull request by hvangeffen**
Closes #11, closes #9

### CODE REVIEW

**Comment by ramcdougal** [comment]
As far as NEURON is concerned, there can be many `finitialize` callbacks, with each one responsible for solving is part of the problem. (e.g. maybe one handles calcium initialization in a certain cell type.) 

Also, there are multiple points in the initialization where the callbacks can be invoked.

**Comment by hvangeffen** [comment]
> As far as NEURON is concerned, there can be many `finitialize` callbacks, with each one responsible for solving is part of the problem. (e.g. maybe one handles calcium initialization in a certain cell type.)

In that case is there a way to dynamically create c function to be passed to hoc install
```c
sym = hoc_install(FINITIALIZE_RULE, function_type, 0, &hoc_top_level_symlist);
sym->u.u_proc->defn.pf = finitialize_callback;
sym->u.u_proc->nauto = 0;
sym->u.u_proc->nobjauto = 0;
```
Or add a callback function accepting an argument.
I would like to be able to do something like:
```c
sym = hoc_install(FINITIALIZE_RULE, function_type, 0, &hoc_top_level_symlist);
sym->u.u_proc->defn.pf = [func_name]() { mexEvalString(func_name + "()"); hoc_ret(); hoc_pushx(0); };
sym->u.u_proc->nauto = 0;
sym->u.u_proc->nobjauto = 0;
```
> Also, there are multiple points in the initialization where the callbacks can be invoked.

As in the same function could be invoked multiple times during the initialization or can you specify when adding the callback when it should be invoked?

File: +neuron/FInitializeHandler.m [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            else
+                type = type_or_func_handle;
+            end
+
+            instance_id = instance_id + 1;
+            self.func_handle = func_handle;
+            self.type = type;
+
+            neuron.FInitializeHandler.handlers(self, instance_id);
+            func_name = "neuron_FInitializeHandler_" + instance_id + "_9b543_hl5345";
```
[the remark this answers is not in the document]
Comment (ramcdougal) [reply +neuron/FInitializeHandler.m:37]: I thought these callbacks were going to be combined with the other callbacks and just have some random UUID or something?

[the remark this answers is not in the document]
Comment (hvangeffen) [reply +neuron/FInitializeHandler.m]: Switched it to use a true UUID

**Pull request merged**